flink-python/pyflink/datastream/checkpoint_storage.py

```diff
@@ -0,0 +1,302 @@
+################################################################################
+#  Licensed to the Apache Software Foundation (ASF) under one
+#  or more contributor license agreements.  See the NOTICE file
+#  distributed with this work for additional information
+#  regarding copyright ownership.  The ASF licenses this file
+#  to you under the Apache License, Version 2.0 (the
+#  "License"); you may not use this file except in compliance
+#  with the License.  You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+# limitations under the License.
+################################################################################
+
+from abc import ABCMeta
+
+from py4j.java_gateway import get_java_class
+
+from pyflink.java_gateway import get_gateway
+
+__all__ = [
+    'CheckpointStorage',
+    'JobManagerCheckpointStorage',
+    'FileSystemCheckpointStorage',
+    'CustomCheckpointStorage'
+]
+
+
+def _from_j_checkpoint_storage(j_checkpoint_storage):
+    if j_checkpoint_storage is None:
+        return None
+    gateway = get_gateway()
+    JCheckpointStorage = gateway.jvm.org.apache.flink.runtime.state.CheckpointStorage
+    JJobManagerCheckpointStorage = gateway.jvm.org.apache.flink.runtime.state.storage \
+        .JobManagerCheckpointStorage
+    JFileSystemCheckpointStorage = gateway.jvm.org.apache.flink.runtime.state.storage \
+        .FileSystemCheckpointStorage
+    j_clz = j_checkpoint_storage.getClass()
+
+    if not get_java_class(JCheckpointStorage).isAssignableFrom(j_clz):
+        raise TypeError("The input %s is not an instance of CheckpointStorage."
+                        % j_checkpoint_storage)
+
+    if get_java_class(JJobManagerCheckpointStorage).isAssignableFrom(j_clz):
+        return JobManagerCheckpointStorage(j_jobmanager_checkpoint_storage=j_checkpoint_storage)
+    elif get_java_class(JFileSystemCheckpointStorage).isAssignableFrom(j_clz):
+        return FileSystemCheckpointStorage(j_fs_checkpoint_storage=j_checkpoint_storage)
+    else:
+        return CustomCheckpointStorage(j_checkpoint_storage)  # users' customized state backend
+
+
+class CheckpointStorage(object):
```

Comment:
```suggestion
class CheckpointStorage(abc.ABC):
```

See https://issues.apache.org/jira/browse/FLINK-19892 for more details.